diff --git a/man/man1/page.1 b/man/man1/page.1
index 8edefe26..cb580e81 100644
--- a/man/man1/page.1
+++ b/man/man1/page.1
@@ -8,7 +8,7 @@ files
 .SH SYNOPSIS
 .B page
 [
-.B -abirPRvVw
+.B -abirPRsvVw
 ]
 [
 .B -p
@@ -78,6 +78,11 @@ option sets the resolution for PostScript and PDF
 files, in pixels per inch.
 The default is 100 ppi.
 The
+.B -s
+option scales the resolution in pixels per inch by the ratio
+of the display's DPI over the given pixels per inch value,
+if using a display with DPI greater than 100.
+The
 .B -r
 option reverses the order in which pages are displayed.
 .PP
diff --git a/src/cmd/page/page.c b/src/cmd/page/page.c
index 24a4b024..1daf1fc2 100644
--- a/src/cmd/page/page.c
+++ b/src/cmd/page/page.c
@@ -81,7 +81,7 @@ afmt(Fmt *fmt)
 void
 usage(void)
 {
-	fprint(2, "usage: page [-biRrwf] [-p ppi] file...\n");
+	fprint(2, "usage: page [-biRrwfs] [-p ppi] file...\n");
 	wexits("usage");
 }
 
@@ -99,6 +99,7 @@ threadmain(int argc, char **argv)
 	enum { Ninput = 16 };
 	uchar buf[Ninput+1];
 	int readstdin;
+	int scaleppi = 0;
 
 	ARGBEGIN{
 	/* "temporary" debugging options */
@@ -145,6 +146,9 @@ threadmain(int argc, char **argv)
 	case 'f':
 		fitwin = 1;
 		break;
+	case 's':
+		scaleppi = 1;
+		break;
 	default:
 		usage();
 	}ARGEND;
@@ -203,7 +207,8 @@ threadmain(int argc, char **argv)
 		wexits("initdraw");
 	}
 	display->locking = 1;
-	ppi = scalesize(display, ppi);
+	if(scaleppi)
+		ppi = scalesize(display, ppi);
 
 	buf[Ninput] = '\0';
 	if(imagemode)
