diff --git a/src/cmd/acme/rows.c b/src/cmd/acme/rows.c
index 7a64fabf..189a30e0 100644
--- a/src/cmd/acme/rows.c
+++ b/src/cmd/acme/rows.c
@@ -316,7 +316,7 @@ rowclean(Row *row)
 void
 rowdump(Row *row, char *file)
 {
-	int i, j, fd, m, n, start, dumped;
+	int i, j, fd, m, n, start, dumped, addscale;
 	uint q0, q1;
 	Biobuf *b;
 	char *buf, *a, *fontname;
@@ -395,8 +395,13 @@ rowdump(Row *row, char *file)
 						goto Continue2;
 				}
 			fontname = "";
+			addscale = 0;
 			if(t->reffont->f != font)
 				fontname = t->reffont->f->name;
+				if(fontname[0] == '/') {
+					fontname = smprint("1*%s", fontname);
+					addscale = 1;
+				}
 			if(t->file->nname)
 				a = runetobyte(t->file->name, t->file->nname);
 			else
@@ -428,6 +433,8 @@ rowdump(Row *row, char *file)
 					100.0*(w->r.min.y-c->r.min.y)/Dy(c->r),
 					w->body.file->b.nc, fontname);
 			}
+			if(addscale)
+				free(fontname);
 			free(a);
 			winctlprint(w, buf, 0);
 			Bwrite(b, buf, strlen(buf));
