diff --git a/src/cmd/devdraw/x11-screen.c b/src/cmd/devdraw/x11-screen.c
index f4aa1ce9..81efa02b 100644
--- a/src/cmd/devdraw/x11-screen.c
+++ b/src/cmd/devdraw/x11-screen.c
@@ -477,6 +477,7 @@ runxevent(XEvent *xev)
 				if(c & Mod1Mask)
 					_x.kbuttons |= 4;
 				if(c & ShiftMask)
+					_x.kbuttons |= 1;
 					shift = 5;
 				gfx_mousetrack(w->client, m.xy.x, m.xy.y, (m.buttons|_x.kbuttons)<<shift, m.msec);
 			}
